File: Python/repositories/UserRepository.py
from sqlalchemy.orm import Session
from models.User import User


class UserRepository:
    """
    Repository SQLAlchemy pour la gestion des utilisateurs.
    Fournit des méthodes CRUD pour manipuler les objets User en base.
    """
    def __init__(self, db: Session):
        """
        Initialise le repository avec une session SQLAlchemy.
        :param db: Session SQLAlchemy
        """
        self.db = db

    def create(self, email: str, password: str, address: str = None) -> User:
        """
        Crée un nouvel utilisateur et le persiste en base.
        :param email: Email de l'utilisateur
        :param password: Mot de passe hashé
        :param address: Adresse postale (optionnelle)
        :return: L'utilisateur créé
        """
        user = User(
            email=email,
            password_hash=password,
            address=address
        )
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user

    def get_by_id(self, user_id: int) -> User:
        """
        Récupère un utilisateur par son identifiant unique.
        :param user_id: Identifiant utilisateur
        :return: L'utilisateur ou None
        """
        return self.db.query(User).filter(User.user_id == user_id).first()

    def get_by_email(self, email: str) -> User:
        """
        Récupère un utilisateur par son email.
        :param email: Email utilisateur
        :return: L'utilisateur ou None
        """
        return self.db.query(User).filter(User.email == email).first()

    def get_all(self) -> list[User]:
        """
        Récupère tous les utilisateurs présents en base.
        :return: Liste d'utilisateurs
        """
        return self.db.query(User).all()

    def update(self, user_id: int, email: str = None, password: str = None, address: str = None) -> User:
        """
        Met à jour les informations d'un utilisateur existant.
        :param user_id: Identifiant utilisateur
        :param email: Nouvel email (optionnel)
        :param password: Nouveau mot de passe hashé (optionnel)
        :param address: Nouvelle adresse (optionnelle)
        :return: L'utilisateur mis à jour ou None
        """
        user = self.get_by_id(user_id)
        if user:
            if email is not None:
                user.email = email
            if password is not None:
                user.password_hash = password
            if address is not None:
                user.address = address
            self.db.commit()
            self.db.refresh(user)
        return user

    def delete(self, user_id: int) -> bool:
        """
        Supprime un utilisateur de la base.
        :param user_id: Identifiant utilisateur
        :return: True si supprimé, False sinon
        """
        user = self.get_by_id(user_id)
        if user:
            self.db.delete(user)
            self.db.commit()
            return True
        return False

File: NetCoreAPI/Repositories/UserRepository.cs
using Microsoft.EntityFrameworkCore;
using NetCoreAPI.Models;
using NetCoreAPI.Data;

namespace NetCoreAPI.Repositories;

/// <summary>
/// Repository pour la gestion des utilisateurs (accès base de données).
/// </summary>
public class UserRepository : IUserRepository
{
    private readonly GestionFormationContext _context;

    /// <summary>
    /// Initialise une nouvelle instance du repository utilisateur.
    /// </summary>
    /// <param name="context">Le contexte de base de données.</param>
    public UserRepository(GestionFormationContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Récupère tous les utilisateurs de la base de données.
    /// </summary>
    /// <returns>Une liste d'utilisateurs.</returns>
    public async Task<IEnumerable<User>> GetAllAsync()
    {
        return await _context.Users.ToListAsync();
    }

    /// <summary>
    /// Récupère un utilisateur par son identifiant.
    /// </summary>
    /// <param name="id">L'identifiant de l'utilisateur.</param>
    /// <returns>L'utilisateur ou null s'il n'existe pas.</returns>
    public async Task<User?> GetByIdAsync(int id)
    {
        return await _context.Users.FirstOrDefaultAsync(u => u.UserId == id);
    }

    /// <summary>
    /// Récupère un utilisateur par son email.
    /// </summary>
    /// <param name="email">L'email de l'utilisateur.</param>
    /// <returns>L'utilisateur ou null s'il n'existe pas.</returns>
    public async Task<User?> GetByEmailAsync(string email)
    {
        return await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
    }

    /// <summary>
    /// Crée un nouvel utilisateur.
    /// </summary>
    /// <param name="user">L'utilisateur à créer.</param>
    /// <returns>L'utilisateur créé.</returns>
    public async Task<User> CreateAsync(User user)
    {
        _context.Users.Add(user);
        await _context.SaveChangesAsync();
        return user;
    }

    /// <summary>
    /// Met à jour un utilisateur existant.
    /// </summary>
    /// <param name="user">L'utilisateur à mettre à jour.</param>
    /// <returns>L'utilisateur mis à jour.</returns>
    public async Task<User> UpdateAsync(User user)
    {
        _context.Users.Update(user);
        await _context.SaveChangesAsync();
        return user;
    }

    /// <summary>
    /// Supprime un utilisateur par son identifiant.
    /// </summary>
    /// <param name="id">L'identifiant de l'utilisateur à supprimer.</param>
    /// <returns>True si l'utilisateur a été supprimé, sinon false.</returns>
    public async Task<bool> DeleteAsync(int id)
    {
        var user = await _context.Users.FirstOrDefaultAsync(u => u.UserId == id);
        if (user == null)
            return false;

        _context.Users.Remove(user);
        await _context.SaveChangesAsync();
        return true;
    }

    /// <summary>
    /// Vérifie si un utilisateur existe par son identifiant.
    /// </summary>
    /// <param name="id">L'identifiant de l'utilisateur.</param>
    /// <returns>True si l'utilisateur existe, sinon false.</returns>
    public async Task<bool> ExistsAsync(int id)
    {
        return await _context.Users.AnyAsync(u => u.UserId == id);
    }

    /// <summary>
    /// Vérifie si un utilisateur existe par son email.
    /// </summary>
    /// <param name="email">L'email à vérifier.</param>
    /// <returns>True si l'email existe, sinon false.</returns>
    public async Task<bool> EmailExistsAsync(string email)
    {
        return await _context.Users.AnyAsync(u => u.Email == email);
    }
}